### PR TITLE
Qt: More options in the tools menu for updating workspaces

### PR DESCRIPTION
**Pull request by petersteneteg**
image

### CODE REVIEW

**Comment by ResearchDaniel** [comment]
How does loading of multiple workspaces work?

File: src/core/network/workspacemanager.cpp [hunk 1]
```diff
@@ -169,22 +169,22 @@ void WorkspaceManager::load(std::istream& stream, const std::string& refPath,
     deserializers_.invoke(deserializer, exceptionHandler);
 }
 
-void WorkspaceManager::save(const std::string& path, const ExceptionHandler& exceptionHandler,
+void WorkspaceManager::save(std::string_view path, const ExceptionHandler& exceptionHandler,
                             WorkspaceSaveMode mode) {
-    auto ostream = filesystem::ofstream(path);
+    auto ostream = filesystem::ofstream(std::string(path));
```
Comment (martinfalk) [inline src/core/network/workspacemanager.cpp:174]: maybe we should provide an override for string_view for both ofstream and ifstream

File: src/core/network/workspaceutils.cpp [hunk 2]
```diff
@@ -36,26 +36,34 @@
...
-void updateWorkspaces(InviwoApplication* app) {
-    updateWorkspaces(app, filesystem::getPath(PathType::Workspaces));
-
-    for (const auto& m : app->getModules()) {
-        updateWorkspaces(app, m->getPath(ModulePath::Workspaces));
-    }
-}
+void forEachWorkspaceInDirRecusive(std::string_view path,
+                                   std::function<void(std::string_view)> callback) {
+    std::function<void(const std::string&)> traverse = [&](const std::string& path) {
```
Comment (martinfalk) [inline src/core/network/workspaceutils.cpp]: why not using `filesystem::getDirectoryContentsRecursively()` here?

File: src/qt/editor/toolsmenu.cpp [hunk 3]
```diff
@@ -181,6 +183,78 @@ ToolsMenu::ToolsMenu(InviwoMainWindow *win) : QMenu(tr("&Tools"), win) {
     auto sourceMenu = addMenu("Create &Sources");
     addInviwoMetaAction(sourceMenu);
 #endif
-}
+
+    auto workspaceMenu = addMenu("Workspaces");
+
+    auto loadExampleWorkspaces = workspaceMenu->addAction("Load Example Workspaces");
```
Comment (martinfalk) [inline src/qt/editor/toolsmenu.cpp:190]: could add a tooltip to those actions explaining the difference between load (dryrun::yes) and update (dryrun::no)

File: src/qt/editor/toolsmenu.cpp [hunk 4]
```diff
@@ -181,6 +183,78 @@ ToolsMenu::ToolsMenu(InviwoMainWindow *win) : QMenu(tr("&Tools"), win) {
...
+    connect(updateRegressionWorkspaces, &QAction::triggered, [win]() {
+        try {
+            util::updateRegressionWorkspaces(win->getInviwoApplication(), util::DryRun::No);
+        } catch (const Exception& e) {
+            util::log(e.getContext(), e.getMessage(), LogLevel::Error);
+        }
+    });
+
+    auto loadWorkspaces = workspaceMenu->addAction("Load Workspaces In Folder...");
+    connect(loadWorkspaces, &QAction::triggered, [win]() {
```
Comment (martinfalk) [inline src/qt/editor/toolsmenu.cpp:241]: is this also recursive?

**Review by martinfalk — CHANGES_REQUESTED** [review]

**Comment by martinfalk** [comment]
> How does loading of multiple workspaces work?

also, what happens if loading a workspace is not successful?

**Comment by petersteneteg** [comment]
Load just opens them one after an other... update opens and saves them again useful for updating the workspace versions

File: src/core/network/workspacemanager.cpp (hunk 1 above)
Comment (petersteneteg) [reply src/core/network/workspacemanager.cpp:174]: problem it that ifstream needs a string or ah fs::path. hence Im sticking with string until we can switch to fs::path

File: src/core/network/workspaceutils.cpp (hunk 2 above)
Comment (petersteneteg) [reply src/core/network/workspaceutils.cpp]: guess so... the function is old I just moved it

File: src/qt/editor/toolsmenu.cpp (hunk 4 above)
Comment (petersteneteg) [reply src/qt/editor/toolsmenu.cpp:241]: yes

File: src/qt/editor/toolsmenu.cpp (hunk 3 above)
Comment (petersteneteg) [reply src/qt/editor/toolsmenu.cpp:190]: mmm

File: include/inviwo/core/network/workspaceutils.h [hunk 5]
```diff
@@ -31,19 +31,25 @@
...
 
-void IVW_CORE_API updateWorkspaces(InviwoApplication* app);
+IVW_CORE_API void forEachWorkspaceInDirRecusive(std::string_view path,
+                                                std::function<void(std::string_view)> callback);
 
-void IVW_CORE_API updateRegressionWorkspaces(InviwoApplication* app);
+enum class DryRun { Yes, No };
 
-void IVW_CORE_API updateWorkspaces(InviwoApplication* app, const std::string& path);
+IVW_CORE_API void updateWorkspaces(InviwoApplication* app, std::string_view path, DryRun dryRun);
```
Comment (ResearchDaniel) [inline include/inviwo/core/network/workspaceutils.h:48]: Explain what DryRun means in this context.
I guess it just opens the workspaces, updates but does not save them?

Comment (ResearchDaniel) [reply include/inviwo/core/network/workspaceutils.h:48]: Probably also want to mention that update refers to the workspace version.
I know that this is more or less updating existing code but we could mention what happens if there are errors while loading a workspace. Is it still updated and saved?

**Review by ResearchDaniel — DISMISSED** [review]
It might be clearer with a slightly different menu structure and naming (order according to usage frequency):
**Workspace version** 
- Update Examples
- Update Regressions
- Update Folder
- Load Examples
- Load Regressions
- Load Folder

**Review by martinfalk — DISMISSED** [review]

**Review by martinfalk — APPROVED** [review]

**Pull request merged**